### PR TITLE
feat: add browser-level PWA install/launch/uninstall APIs

### PR DESCRIPTION
**Pull request by LuHuangMSFT**
## What / Why

Adds first-class **Progressive Web App (PWA) automation** to the `Browser` API, wrapping Chromium's browser-scoped `PWA.*` CDP domain so callers can install, launch, inspect, and uninstall PWAs without managing raw sessions.

Implements the API agreed on in #15202.

## API

```ts
browser.installPWA({manifestId, installUrlOrBundleUrl, displayMode?}) // => manifestId
browser.uninstallPWA({manifestId})                                    // => void
browser.launchPWA({manifestId, url?, timeout?})                       // => Page
browser.getPWAState({manifestId})                                     // => {badgeCount, fileHandlers}
```

## Behavior and compatibility

- Chromium/CDP only; WebDriver BiDi implementations throw `UnsupportedOperation`.
- The `PWA` domain requires an opt-in pipe connection (`puppeteer.launch({pipe: true})`) and is unavailable over WebSocket.
- `installUrlOrBundleUrl` is required because the browser-scoped session has no associated page from which Chromium could derive the install URL.
- `launchPWA` uses the tab target id returned by `PWA.launch` to resolve its child page target. Existing app windows that Chromium focuses resolve to their existing `Page`.
- `timeout` controls how long `launchPWA` waits for the app's page target. The protocol command uses the configured `protocolTimeout`; `0` disables the target-wait timeout.

## Testing

The PWA integration suite covers lifecycle/state, default and explicit-URL launch, and standalone display mode. Unsupported Firefox, BiDi, BiDi-only, and headless-shell configurations are expected-skipped. Focused unit coverage verifies that the launch timeout is applied to target waiting without overriding the protocol timeout.

Validated with Chrome for Testing 150 and a Chromium/Edge build in headless and headful modes. The exact PR build was also consumed by a local `chrome-devtools-mcp` PWA category and passed a real stdio MCP lifecycle: install, inspect OS state, launch/select the returned `Page`, execute a page script in standalone mode, uninstall, and verify state rejection. Generated docs, API extraction, formatting, and lint checks pass.

Refs #15202

### CODE REVIEW

File: packages/puppeteer-core/src/cdp/Browser.ts [hunk 1]
```diff
@@ -523,6 +528,103 @@ export class CdpBrowser extends BrowserBase {
...
+    const knownPageTargetIds = new Set(
+      this.targets()
+        .filter(target => {
+          return target.type() === 'page';
+        })
+        .map(target => {
+          return (target as CdpTarget)._targetId;
+        }),
+    );
+    await this.#connection.send('PWA.launch', {
```
Comment (OrKoN) [inline packages/puppeteer-core/src/cdp/Browser.ts]: Can we use the targetId that this command returns to find the page target?

Comment (LuHuangMSFT) [reply packages/puppeteer-core/src/cdp/Browser.ts]: `PWA.launch` resolves with the launched **tab** target's id (the CDP `PWA.LaunchResponse` docs describe it as "ID of the tab target created as a result"), not the page. Tab targets aren't exposed through `browser.targets()`, so awaiting the returned id directly never resolves, and the web contents are a separate `page` target.

So we can use it, just as a correlation key rather than a direct lookup. I just changed it to look up that tab target in the target manager and wait for the `page` target that the manager links to it as a child. That drops the origin-based matching entirely and is deterministic for same-origin and multi-window launches. Verified on Chrome for Testing 150 and a Chromium build, headless and headful.

**Comment by LuHuangMSFT** [comment]
@googlebot I signed it!

**Comment by LuHuangMSFT** [comment]
@googlebot I signed it!

**Comment by OrKoN** [comment]
Thanks, we need to update TestExpectations.json file to specify that failures on Firefox, chrome-headless-shell and chrome-bidi are expected.

File: packages/puppeteer-core/src/cdp/Browser.ts (hunk 1 above)
Comment (OrKoN) [reply packages/puppeteer-core/src/cdp/Browser.ts]: Thanks, the current version looks good to me.

**Comment by LuHuangMSFT** [comment]
Done in f833bad. Added `[pwa.test] *` skip entries for `firefox`, `webDriverBiDi`, and `chrome` + `chrome-headless-shell`, mirroring the existing `extensions.test` expectations. Verified locally that the suite still runs and passes on chrome-pipe/headless/headful and is skipped on the three unsupported configs.

**Comment by LuHuangMSFT** [comment]
@OrKoN The follow-up fixes are now complete through 1fed900. The install URL contract and launch timeout match runtime behavior, BiDi-only is expected-skipped, and launch cleanup now covers late responses, timeout-disabled waits, browser disconnect, and page creation. I also built the exact PR into a local chrome-devtools-mcp PWA category and passed a real stdio MCP lifecycle: install, inspect OS state, launch/select the returned Page, execute a page script in standalone mode, uninstall, and verify state rejection. The PR is ready for re-review; the new full CI run will need maintainer approval to start.

File: packages/puppeteer-core/src/cdp/Browser.ts [hunk 2]
```diff
@@ -523,6 +529,99 @@ export class CdpBrowser extends BrowserBase {
...
+          // the CDP `PWA.LaunchResponse` docs). Tab targets sit above page
+          // targets in the target hierarchy and are not exposed through
+          // `browser.targets()`, so the returned id can't be awaited directly.
+          const {targetId: tabTargetId} = await this.#connection.send(
+            'PWA.launch',
+            {
+              manifestId: options.manifestId,
+              url: options.url,
+            },
+            {timeout},
```
Comment (OrKoN) [inline packages/puppeteer-core/src/cdp/Browser.ts]: I think we should let this run with the default timeout.

File: packages/puppeteer-core/src/cdp/Browser.ts [hunk 3]
```diff
@@ -523,6 +529,99 @@ export class CdpBrowser extends BrowserBase {
...
+                return false;
+              }
+              for (const child of tab._childTargets()) {
+                if (child === candidate) {
+                  return true;
+                }
+              }
+              return false;
+            },
+            {timeout: 0, signal: controller.signal},
```
Comment (OrKoN) [inline packages/puppeteer-core/src/cdp/Browser.ts]: instead we should pass timeout here.

File: packages/puppeteer-core/src/cdp/Browser.ts [hunk 4]
```diff
@@ -523,6 +529,99 @@ export class CdpBrowser extends BrowserBase {
...
+  }
+
+  override async launchPWA(options: LaunchPWAOptions): Promise<Page> {
+    const timeout = options.timeout ?? 30_000;
+    const controller = new AbortController();
+    const timeoutDeferred = Deferred.create<Page>({
+      timeout,
+      message: `Timed out after waiting ${timeout}ms`,
+    });
+    const disconnectedDeferred = Deferred.create<Page, TargetCloseError>();
```
Comment (OrKoN) [inline packages/puppeteer-core/src/cdp/Browser.ts]: I don't think we need to handle this extra. If the browser is disconnected, the CDP command will error. The waitForTarget should just time out according to the timeout. What was the reason to add handling of the disconnect event?

**Review by OrKoN — CHANGES_REQUESTED** [review]
Thanks! A few comments:

File: packages/puppeteer-core/src/cdp/Browser.ts (hunk 4 above)
Comment (LuHuangMSFT) [reply packages/puppeteer-core/src/cdp/Browser.ts]: The intent was to cover the timeout: 0 case. After PWA.launch resolves, there is no pending CDP command, and waitForTarget does not observe browser disconnection, so it could remain pending indefinitely and retain its listeners. However, that adds custom behavior beyond the normal waitForTarget semantics. I agree consistency is preferable here, so I’ll remove the disconnect race and rely on the configured target-wait timeout.

File: packages/puppeteer-core/src/cdp/Browser.ts (hunk 2 above)
Comment (LuHuangMSFT) [reply packages/puppeteer-core/src/cdp/Browser.ts]: I’ll remove the per-command timeout so PWA.launch uses the configured protocolTimeout.

File: packages/puppeteer-core/src/cdp/Browser.ts (hunk 3 above)
Comment (LuHuangMSFT) [reply packages/puppeteer-core/src/cdp/Browser.ts]: I’ll pass options.timeout directly to waitForTarget, allowing its normal 30-second default when unspecified and 0 to disable the wait timeout.

File: packages/puppeteer-core/src/cdp/Browser.ts (hunk 2 above)
Comment (LuHuangMSFT) [reply packages/puppeteer-core/src/cdp/Browser.ts]: Addressed in e9a7e14. `PWA.launch` now uses the configured protocol timeout.

File: packages/puppeteer-core/src/cdp/Browser.ts (hunk 3 above)
Comment (LuHuangMSFT) [reply packages/puppeteer-core/src/cdp/Browser.ts]: Addressed in e9a7e14. `options.timeout` is now passed directly to `waitForTarget`.

File: packages/puppeteer-core/src/cdp/Browser.ts (hunk 4 above)
Comment (LuHuangMSFT) [reply packages/puppeteer-core/src/cdp/Browser.ts]: Addressed in e9a7e14. I removed the explicit disconnect race, so `launchPWA` now follows the normal `waitForTarget` semantics.

**Comment by LuHuangMSFT** [comment]
@OrKoN Addressed the three `launchPWA` comments in e9a7e14: `PWA.launch` now uses the configured protocol timeout, `options.timeout` is passed to `waitForTarget`, and the explicit disconnect race is removed. I also updated the focused test, public docs, and PR description, then revalidated Edge/Chrome and the MCP lifecycle. Ready for re-review.

File: test/src/cdp/pwa.test.ts [hunk 5]
```diff
@@ -0,0 +1,105 @@
+/**
+ * @license
+ * Copyright 2025 Google Inc.
```
Comment (OrKoN) [inline test/src/cdp/pwa.test.ts]:
```suggestion
 * Copyright 2026 Google Inc.
```

**Review by OrKoN — APPROVED** [review]
LGTM thanks
